types/react-router/index.d.ts

```diff
@@ -99,7 +99,16 @@ export interface match<P> {
   url: string;
 }
 
+// Diff / Omit taken from https://github.com/Microsoft/TypeScript/issues/12215#issuecomment-311923766
+export type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];
+export type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;
+
 export function matchPath<P>(pathname: string, props: RouteProps): match<P> | null;
-export function withRouter<P>(component: React.ComponentType<RouteComponentProps<any> & P>): React.ComponentClass<P>;
+export function withRouter<P extends RouteComponentProps<any>>(component: React.ComponentType<P>): React.ComponentClass<Omit<P, keyof RouteComponentProps<any>>>;
```

Comment:
I have ran into an issue with this change and would be interested to get your thoughts @Grmiade: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/24070

Separately, do you know why the old type definition didn't work? It seems like it should. If you provide the generic explicitly, it works as expected. Perhaps it's a bug with TypeScript's generic inference? I filed a TypeScript bug here: https://github.com/Microsoft/TypeScript/issues/21964

Reply:
We wanted to fix this issue https://github.com/DefinitelyTyped/DefinitelyTyped/issues/17181
And make the function `withRouter` smarter. No need to provide explicit type. It can be inferred.
The typings becomes more safe. But indeed, you found a bug with the `Omit` definition.

Reply:
We were also hitting https://github.com/DefinitelyTyped/DefinitelyTyped/issues/17181. Thanks for fixing it!

I'm still curious why the old definition didn't work. I would have expected the return type to omit the router props, because the generic excludes the router props. Is that what you would expected? However, it seems TypeScript is not inferring the generic to exclude the router props. I filed a TS bug for that here—please chime in! https://github.com/Microsoft/TypeScript/issues/21964.

Reply:
```typescript
interface TOwnProps extends RouteComponentProps<any> {
    username: string;
}

const ComponentFunction = (props: TOwnProps) => (
    <h2>Welcome {props.username} on {props.location.pathname}</h2>
);

const WithRouterComponentFunction = withRouter(ComponentFunction);
// withRouter<TOwnProps>(component: React.ComponentType<RouteComponentProps<any> & TOwnProps>): React.ComponentClass<TOwnProps>
```

 #17181 issue appears when `TOwnProps ` already includes RouteComponentProps because the returned component includes RouteComponentProps too.

Indeed Typescript is not inferring the generic to exclude the router props. Bug or miss-feature, I don't know :/